content/blog/2018-02-07-update-on-async-rendering.md

```diff
@@ -0,0 +1,123 @@
+
+The upgrade path for this is to move data-fetching into `componentDidMount`:
+`embed:update-on-async-rendering/fetching-external-data-after.js`
+
+> Note:
+>
+> Some advanced use-cases (e.g. libraries like Relay) may want to experiment with eagerly prefetching async data. An example of how this can be done is available [here](https://gist.github.com/bvaughn/89700e525ff423a75ffb63b1b1e30a8f).
+
+### Adding event listeners (or subscriptions)
+
+Here is an example of a component that subscribes to an external event dispatcher when mounting:
+`embed:update-on-async-rendering/adding-event-listeners-before.js`
+
+Unfortunately, this can cause memory leaks for server rendering (where `componentWillUnmount` will never be called) and async rendering (where rendering might be interrupted before it completes, causing `componentWillUnmount` not to be called).
+
+People often assume that `componentWillMount` and `componentWillUnmount` are paired, but that is not guaranteed. Only once `componentDidMount` has been called does React guarantee that `componentWillUnmount` will later be called for clean up.
+
+For this reason, the recommended way to add listeners/subscriptions is to use the `componentDidMount` lifecycle:
+`embed:update-on-async-rendering/adding-event-listeners-after.js`
+
+### Updating `state` based on `props`
+
+Here is an example of a component that uses the legacy `componentWillReceiveProps` lifecycle to update `state` based on new `props` values:
+`embed:update-on-async-rendering/updating-state-from-props-before.js`
+
+Although the above code is not problematic in itself, the `componentWillReceiveProps` lifecycle is often mis-used in ways that _do_ present problems. Because of this, the method has been deprecated.
+
+As of version 16.3, the recommended way to update `state` in response to `props` changes is using the new `static getDerivedStateFromProps` lifecycle:
+`embed:update-on-async-rendering/updating-state-from-props-after.js`
+
+> Note:
+>
+> The [`react-lifecycles-compat`](https://github.com/reactjs/react-lifecycles-compat) polyfill allows this new lifecycle to be used with older versions of React as well. This can be helpful if you're writing a shared component that is intended for use with multiple versions of React.
+
+### Invoking external callbacks
+
+Here is an example of a component that calls an external function when its internal state changes:
+`embed:update-on-async-rendering/invoking-external-callbacks-before.js`
+
+This would not be safe to do in async mode, because the external callback might get called multiple times for a single update. Instead, the `componentDidUpdate` lifecycle should be used since it is guaranteed to be invoked only once per update:
+`embed:update-on-async-rendering/invoking-external-callbacks-after.js`
+
+## Open source project maintainers
```

Comment:
I'd add a section before this one. Something like

>## Other scenarios

>While we tried to cover the most common use cases in this post, we recognize that we might have missed some of them. If you are using `componentWillMount`, `componentWillUpdate`, or `componentWillReceiveProps` in ways that aren't covered by this blog post, and aren't sure how to migrate off these legacy lifecycles, please [file a new issue against our documentation](https://github.com/reactjs/reactjs.org/issues/new) with your code examples and as much background information as you can provide. We will update this document with new alternative patterns as they come up.

Reply:
Love it.